Plot plugin support

langers2k opened the pull request:
I did some work on plot plugin about 6 months ago, as that went stale I'm starting the PR process again. Following the comments last time, this is probably the smallest subset of code that will actually allow a user to show a plot.

I'm not sure how I can make this more painless so any comments are welcome.

The following will allow a user to create a line plot from a properties file without setting any of the additional (optional) fields. It's written in a way that should allow clean changes on top to add the additional functionally required to get full control of the plot plugin.

Example:
```
plotplugin {
  line('graph_data.csv', 'My plots') {
    propertiesFile {
      file 'data.prop'
    }
  }
}
```

langers2k commented:
I also don't know why it's failing to build...

Builds here with no issue :/

daspilker commented on job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/publisher/PlotPluginContext.groovy:
```diff
@@ -0,0 +1,19 @@
+import javaposse.jobdsl.dsl.ContextHelper
+import javaposse.jobdsl.dsl.Context
+
+class PlotPluginContext implements Context {
```
Rename this class to PlotContext, the other contexts also do not contain "plugin" in their name.

daspilker commented on job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/publisher/PlotPluginContext.groovy:
```diff
@@ -0,0 +1,19 @@
+class PlotPluginContext implements Context {
+    List<PlotPluginPlotContext> plots = []
+
+    void line(String dataStore, String group, Closure plotPluginPlotClosure) {
```
All plot styles have the same options, use separate method to set the style, e.g. `style(String style)`. The dataStore option is not visible in the UI, so why must this be specified in the DSL?

In the UI, the group is a mandatory option, so it should be a method parameter, e.g.

```groovy
plot(String group) {
    style(String style)
    ...
}
```

daspilker commented on job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/publisher/PlotPluginPlotSeriesContext.groovy:
```diff
@@ -0,0 +1,23 @@
+
+import javaposse.jobdsl.dsl.Context
+
+class PlotPluginPlotSeriesContext implements Context {
```
Rename this to PlotSeriesContext.

daspilker commented on job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/publisher/PublisherContext.groovy:
```diff
@@ -338,6 +338,77 @@ class PublisherContext implements Context {
+    *     </plots>
+    * </hudson.plugins.plot.PlotPublisher>
+    */
+    void plotPlugin(Closure plotPluginClosure) {
```
Rename this to `plots` or `plotBuildData`.

daspilker commented:
OK, I got lost and need to start over.

This is how the DSL should look like:

```groovy
publishers {
    plotBuildData {   // PlotsContext
        plot(String group) {    // PlotContext
            style(String style)
            dataSeries { // PlotDataSeriesContext
                propertiesFile(String fileName)
            }
        }
    }
}

langers2k commented:
The plotplugin relies on a CSV datastore to manage it's data, this is normally created by plotplugin and has a random name. I saw a few of choices here:

- Manually set the name with the DSL (with some extra documentation explaining what it is),
- Reuse the code or fucnction from plotplugin to create the random filename
- Use a constant value

Out of these, the first seemed the most sensible to ensure we don't have any conflicts within the workspace that the user can't resolve. I'm happy to try any of those three or even allow the user to specify a value with a default value (probably random) for those that don't.

I'll rework things to give more sensible names and an input style similar to the one above.

langers2k commented:
Updated as per the first set of comments.

I've not added the dataSeries level as the diff was getting messy with the file renames. Assuming this is better I'll sort out the possible data series next.

I've also added a check for all the chart styles and made line the default.

daspilker commented:
OK, great. I also think that manually choosing the datastore file name is the best option along with some documentation.

Add tests and documentation. I would suggest to add the other data series on later pull requests to get the basic features done first.

langers2k commented:
Updated, had to make a small fix and decided to make fileName a required part of the propertiesFile as it shouldn't be option.

Let me know if you want me to flatten the four code commits or even all six.

Thanks,

daspilker commented:
OK, thanks! I squashed the commits and merged them manually.
